codechecker_common/checker_labels.py

```diff
@@ -0,0 +1,207 @@
+                     By default all analyzers are searched.
+        """
+        collection = []
+
+        filter_labels = set(map(self.__get_label_key_value, filter_labels))
+
+        for analyzer, checkers in self.__data.items():
+            if analyzers is None or analyzer in analyzers:
+                for checker, labels in checkers.items():
+                    labels = set(map(self.__get_label_key_value, labels))
+
+                    if labels.intersection(filter_labels):
+                        collection.append(checker)
+
+        return collection
+
+    def label_of_checker(self, checker, label, analyzers=None):
+        """
+        If a label has unique constraint then this function retuns the value
+        that belongs to the given label or the default value that is set among
+        label constraints. If there is no unique constraint then an iterable
+        object returns with the values assigned to the given label. If the
+        checker name is not found in the label config file then its prefixes
+        are also searched. For example "clang-diagnostic" in the config file
+        matches "clang-diagnostic-unused-argument".
+        """
+        labels = chain.from_iterable([])
+
+        for analyzer, _ in self.labels_of_checker(checker, analyzers):
+            if analyzers is None or analyzer in analyzers:
+                labels = chain(labels, filter(
+                    lambda lab: lab[0] == label,
+                    self.labels_of_checker(checker)))
+
+        if label in CheckerLabels.UNIQUE_LABELS:
+            try:
+                return next(labels)[1]
+            except Exception:
+                return CheckerLabels.UNIQUE_LABELS[label]
+
+        return map(lambda _, value: value, labels)
+
+    def severity(self, checker, analyzers=None):
+        """
+        Shorthand for the following call:
+        checker_labels.label_of_checker(checker, 'severity', analyzers)
+        """
+        return self.label_of_checker(checker, 'severity', analyzers)
+
+    def labels_of_checker(self, checker, analyzers=None):
+        """
+        Return the list of labels of a checker. The list contains (label,
+        value) pairs. If the checker name is not found in the label config file
+        then its prefixes are also searched. For example "clang-diagnostic" in
+        the config file matches "clang-diagnostic-unused-argument".
+        """
+        labels = []
+
+        for analyzer, checkers in self.__data.items():
+            if analyzers is None or analyzer in analyzers:
+                c = checker
+
+                if c not in checkers:
+                    c = next(filter(
+                        lambda c: checker.startswith(c),
+                        iter(checkers.keys())), None)
+
+                labels.extend(
+                    map(self.__get_label_key_value, checkers.get(c, [])))
+
+        return labels
+
+    def get_description(self, label):
```

Comment:
I was thinking of this, if it's just a question of style or does it have any pros/cons? I found that by transforming it to a property leaks the data representation, so the user of that property needs to know the structure of the underlying multilevel dictionary. On the other hand if we remove the function's parameter, it becomes less verbose and carries less information about what key needs to be given to the function. So I found these reasons supporting the current implementation. I can't see the advantages of a property yet, but if you are aware of any, please, share it. I think it would be a relevant question in the future too, because we already use a lot of properties at other places and I'm not sure if they're reasonable. It would be interesting and useful to define some rule of thumb about when to use properties.